File: MachineProblems/Code/MP3.java
import java.util.Scanner;

public class MP3 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        String parenthesis = sc.nextLine();

        while (parenthesis.trim().isEmpty()) {
            System.out.println("PLEASE ENTER A VALID STRING");
            parenthesis = sc.nextLine();
        }

        char[] stack = new char[parenthesis.length()];
        int top = -1;

        for (char ch : parenthesis.toCharArray()) {

            if (Character.isWhitespace(ch)) {
                continue;
            }

            if ("()[]{}".indexOf(ch) == -1) {
                System.out.println("NOT BALANCED");
                return;
            }

            if (ch == '(' || ch == '[' || ch == '{') {
                stack[++top] = ch; 
            } 
            else { 

                if (top == -1) {
                    System.out.println("NOT BALANCED");
                    return;
                }

                char last = stack[top--];

                if ((ch == ')' && last != '(') ||
                    (ch == ']' && last != '[') ||
                    (ch == '}' && last != '{')) {
                    System.out.println("NOT BALANCED");
                    return;
                }
            }
        }

        if (top == -1) {
            System.out.println("BALANCED");
        } else {
            System.out.println("NOT BALANCED");
        }

        sc.close();
    }
}

File: MachineProblems/Code/MP3.py
parenthesis = input()

while parenthesis == "" or parenthesis.isspace():
    print("PLEASE ENTER A VALID STRING")
    parenthesis = input()

stack = []

for char in parenthesis:
    
    if char.isspace():
        continue

    if char not in r'()[]{}':
        print("NOT BALANCED")
        quit()
        

    if char in '([{':
        stack.append(char)
        
    elif char in ')]}':
        
        if not stack:
            print("NOT BALANCED")
            quit()
            break
        
        top = stack.pop()
        
        if (char == ')' and top != '(') \
        or (char == ']' and top != '[') \
        or (char == '}' and top != '{'):
            print("NOT BALANCED")
            quit()

        
if not stack:
    print("BALANCED")
else:
    print("NOT BALANCED")